### ISSUE TITLE
Graceful proxy shutdown

### ISSUE THREAD

**Issue by olix0r** [runconduit/conduit#430]
When the proxy receives a shutdown signal (`SIGINT` or `SIGTERM`), it should:
* Stop accepting new inbound connections
* Stop accepting new inbound requests
* Send GOAWAY on all inbound h2 connections
* Wait for all active inbound & outbound requests to complete, or for some (env-defined) max time to expire.
* Exit the process

**Comment by olix0r** [comment]
This likely needs to be broken into some specific smaller tasks (with testing!)

**Comment by olix0r** [comment]
As we investigate #549, we should begin to figure out a plan for proper shutdown logic.

**Comment by briansmith** [comment]
> Wait for all active inbound & outbound requests to complete, or for some (env-defined) max time to expire.

Kuberenetes already has a pod-level configuration setting for this: `terminationGracePeriodSeconds`. I propose that we don't implement our own deadline, but instead just let Kubernetes send us SIGKILL after `terminationGracePeriodSeconds`.

> Send GOAWAY on all inbound h2 connections

I'm not convinced that it is necessary to send GOAWAY and I think it might be a bad idea, especially if the endpoint we send GOAWAY to doesn't support graceful retries. I think we should consider this more carefully.

In particular, I think stopping the acceptance of new connections, and closing the connections to the other containers in the pod when we don't have any requests pending for them, is enough.

**Comment by briansmith** [comment]
> In particular, I think stopping the acceptance of new connections, and closing the connections to the other containers in the pod when we don't have any requests pending for them, is enough.

...and exiting the process when inbound and outbound request queues are empty, of course.

**Comment by briansmith** [comment]
@olix0r It would be really unfortunate to have the behavior of #581 without fixing this issue too, for 0.4.0, because #581 breaks rolling deployments, and I think we should keep rolling deployments working throughout the series of pre-releases so that people can properly evaluate Conduit. Breaking rolling deployments, especially without retry logic, would be a showstopper in terms of evaluation, IMO.

**Comment by seanmonstar** [comment]
Setting Connection: close headers in h1 and sending a GOAWAY in h2 is straight forward, and dropping the TcpListener to stop accepting new connections.

It's not clear how we can gracefully close opaque TCP streams, other then just hopping they close naturally before Kubernetes sends us SIGKILL, but perhaps there isn't anything to be done there

**Comment by briansmith** [comment]
> It's not clear how we can gracefully close opaque TCP streams, other then just hopping they close naturally before Kubernetes sends us SIGKILL, but perhaps there isn't anything to be done there

It seems like if the proxied service is going to accept an inbound connection after SIGTERM, then the Conduit proxy should accept an inbound connection after SIGTERM. Maybe only if/when the proxied service is shut down after SIGTERM (new connections to it fail after SIGTERM) should refuse inbound connections? It seems like we "just" need to be sure that we don't keep connections open to the proxied service that would cause the proxied service to wait for our connections to finish. And, this seems the same for both HTTP and non-HTTP (TCP) connections.

Thinking about this critically, I think we need to be clearer on the goals here. AFAICT the goals are:

1. Don't keep the proxied service running longer than necessary by keeping open connections to it. Assume the proxied service is waiting for us to close those connections before it exits.
2. Process any/all requests that we receive.
3. Maybe: Let clients know that we don't want more connections/requests when possible (`Connection: close` and/or `GOAWAY`). I'm not sure this is necessary.
4. Exit when we don't have any incoming requests to process; i.e. all request queues are empty.

It's really unclear whether we should actually stop accepting new connections, TCP or HTTP, unless we know the proxied service has stopped accepting new connections. In the Conduit-to-Conduit case, the pod that is the source of the requests will remove the terminating pod from its list of destinations when the service discovery service gets notified that the terminating pod has started to terminate. So, traffic incoming should go away on its own without us having to refuse it.

tl;dr: The problem that #581 tried to resolve was that we were staying alive even when there was no work to do after SIGTERM, but it solved it by not waiting even when we had work to do. If we have work to do, we should do it. When we stop having work to do, we should exit. If the incoming stream of work never ends then Kubernetes will kill us.

WDYT?

**Comment by seanmonstar** [comment]
> It seems like if the proxied service is going to accept an inbound connection after SIGTERM, then the Conduit proxy should accept an inbound connection after SIGTERM.

This greatly complicates things, and also means that apps that don't think to do anything for graceful shutdown can never benefit. If we proactively tell the connections (in HTTP, anyways) that it is closing soon, we can essentially improve shutdown for the proxied app, without the developers having had to do so.

> If we have work to do, we should do it. When we stop having work to do, we should exit.

Yes, at the base level, that's exactly it. We need to do something about the listeners, though, otherwise we can't determine if there is no more work. Unless we decided "There's no active connections, stop listening now".

> If the incoming stream of work never ends then Kubernetes will kill us.

I feel that the point of graceful shutdown is to try to make connections stop wherever a good stopping point is. We can force those stopping points to show up with `Connection: close`, `GOAWAY`, and closing the listeners. If we don't, and hope that the proxied app was designed to shutdown properly, then in cases where it didn't, Kubernetes will eventually SIGKILL everything at possible poor times. It's much worse to die in the middle of a stream (or worse, right at the end when one peer thinks the connection was a success, but the other peer didn't have time to do anything).

**Comment by briansmith** [comment]
> > It seems like if the proxied service is going to accept an inbound connection after SIGTERM, then the Conduit proxy should accept an inbound connection after SIGTERM.

> This greatly complicates things [...]

To be clear, there I'm describing what I see as the ideal semantics. I don't think we can reasonably implement that anyway, like you say, but it's useful to see what the ideal is.

> Unless we decided "There's no active connections, stop listening now".

That's exactly what I'm suggesting. Exit when there are no active connections, but don't stop accepting connections until that point. Would this be hard to do? Obviously this is a race.

> I feel that the point of graceful shutdown is to try to make connections stop wherever a good stopping point is. We can force those stopping points to show up with Connection: close, GOAWAY, and closing the listeners.

Ideally the source of incoming traffic has retry logic so that receiving a `Connection: clone` or refused connection, etc. In that case, I do think it makes sense to be more proactive in refusing traffic.

> If we don't, and hope that the proxied app was designed to shutdown properly, then in cases where it didn't, Kubernetes will eventually SIGKILL everything at possible poor times.

If the proxied app wasn't designed to shut down in a reasonable way then we don't need to shut down any better. In general the proxy just needs to minimize trouble it causes the proxied app by avoiding shutting down too early and avoiding shutting down too late.

> It's much worse to die in the middle of a stream (or worse, right at the end when one peer thinks the connection was a success, but the other peer didn't have time to do anything).

Right. I agree with this.

Here's the issue I see:
1. Some other pods queue up requests (HTTP or otherwise) that are scheduled for our pod.
2. The user rolls the deployment that our pod is part of. As part of rolling the deployment, Kubernetes decides to terminate our pod and sends each container in our pod SIGTERM.

When the traffic from those other pods gets sent to us, IMO it shouldn't matter whether that happens on a new connection or on an existing connection, as far as whether we're going to service that request after SIGTERM. That's why it makes more sense to me to stop accepting connections when all queues are empty, hoping that all such traffic gets sent to us before `terminationGracePeriodSeconds` has elapsed. That is, I think that stopping the acceptance of new connections immediately on SIGTERM is definitely reasonable, but waiting until requests queues are empty seems better to me, if it can be implemented in a reasonable way.

**Comment by seanmonstar** [comment]
> Exit when there are no active connections, but don't stop accepting connections until that point. Would this be hard to do?

It wouldn't be hard technically. I just wonder if it's the best thing we should do. I think that if we stop accepting connections, a client will see the errors and adapt (probably by retrying the request after another DNS lookup).

I *can* see a case where a service is asked to shutdown, and as part of its shutdown, wants to shoot off a last message to something else, like a metrics dump. In that case, perhaps we'd just want to stop listening for **inbound** connections, but not close the **outbound** right away?

> As part of rolling the deployment, Kubernetes decides to terminate our pod and sends each container in our pod SIGTERM.

I think some of our discussion and decision seems to be depending on the exact behavior of Kubernetes. Do we know its exact behavior? When it sends a SIGTERM, will it have already removed the pods ports from the cluster? If it has, then it does seem like we can leave the listener alone.

**Comment by olix0r** [comment]
@briansmith 

> It seems like if the proxied service is going to accept an inbound connection after
> SIGTERM, then the Conduit proxy should accept an inbound connection after SIGTERM.

What are some situations where an application _wants_ to accept new _connections_ after
the scheduler has told it to die.

> Thinking about this critically, I think we need to be clearer on the goals here. AFAICT
> the goals are:

I think we can _do better_ than most applications with regard to draining and lifecycle
management. Experience tells me that the vast majority of applications _do not_ do this
well, and so we have an opportunity to improve on default application behavior.

> - Maybe: Let clients know that we don't want more connections/requests when possible
>   (Connection: close and/or GOAWAY). I'm not sure this is necessary.

I don't think that this is strictly required, though it is surely advantageous.

> In the Conduit-to-Conduit case, the pod that is the source of the requests will remove
> the terminating pod from its list of destinations when the service discovery service
> gets notified that the terminating pod has started to terminate. So, traffic incoming
> should go away on its own without us having to refuse it.

Though, this assumes service discovery notifications are faster than shutdown. I don't
think we have any type of guarantees on this (we certainly do not outside of Kubernetes).
We should try to operate on the assumption that service discovery will lie sometimes, or
at least be a little out of date. I'm quite partial to the belt-and-suspenders approach;
though I think it's fine to prioritize "more graceful" shutdown later than something that
basically works in k8s today.

**Comment by briansmith** [comment]
> What are some situations where an application wants to accept new connections after
the scheduler has told it to die?

When the source of the incoming traffic copes poorly with connections that are refused and helping them is more important than shutting down as fast as possible.

> I think we can *do better* than most applications with regard to draining and lifecycle
> management. Experience tells me that the vast majority of applications do not do this
> well, and so we have an opportunity to improve on default application behavior.

I agree. However, I think there's no one definition of what's actually better/best.

If I understand linkerd's code correctly, linkerd waits for its equivalent of `terminationGracePeriodSeconds` and then exits, and doesn't do anything to stop incoming traffic or tell it to go away:
https://github.com/linkerd/linkerd/blob/863fc938a1dc2091e5b7045add35da8ac7e50b5a/linkerd/main/src/main/scala/io/buoyant/linkerd/Main.scala#L151-L162

Regardless, I think we can implement the change that we all agree on, namely draining the request queues. I am fine with using the "stop accepting new connections right away" method now (We already stop accepting new incoming connections right away since we exit right away, IIUC, so this wouldn't even be a change in behavior.) Then we can see if we need to make the behavior configurable and/or change the default behavior in some way later.

**Comment by seanmonstar** [comment]
> If I understand linkerd's code correctly, linkerd waits for its equivalent of terminationGracePeriodSeconds and then exits, and doesn't do anything to stop incoming traffic or tell it to go away:

It was hard to find for me too, but it is actually telling everything to close: https://github.com/linkerd/linkerd/blob/863fc938a1dc2091e5b7045add35da8ac7e50b5a/linkerd/main/src/main/scala/io/buoyant/linkerd/Main.scala#L47-L51

That in turn closes all the various finagle `Service`s, which I'm sure does some magic.

**Comment by briansmith** [comment]
> It was hard to find for me too, but it is actually telling everything to close

Gotcha. It is telling everything to close with a deadline of the given duration, not waiting for the given duration and then telling everything to close.

**Comment by olix0r** [comment]
> It is telling everything to close with a deadline of the given duration, not waiting for the given duration and then telling everything to close.

Yes, this is a preferable API choice, if we had some way of discovering the k8s policy and informing the proxy of it.

**Comment by slumbi** [comment]
In our Kubernetes cluster we have a service using sticky sessions , with a high `terminationGracePeriodSeconds` of 5minutes. On shutdown the service becomes not-ready, and does not receive traffic from new users anymore, but is still able to serve existing sessions for up to N(=5) minutes.
Our issue is that the linkerd proxy sidecar does not respect `terminationGracePeriodSeconds` and terminates before our service. I did not didn't find  any configuration  to change this.  (stable-2.3.2)

**Comment by grampelberg** [comment]
@slumbi please open a separate issue as I think this’ll require some extra discussion.
